modules/frontend/searchsharding.go

```diff
@@ -232,41 +143,49 @@ func (s searchSharder) RoundTrip(r *http.Request) (*http.Response, error) {
 	}
 	overallResponse.resultsMetrics.TotalBlockBytes = totalBlockBytes
 
+	// TODO: maybe it's better to loop on overallResponse.shouldQuit()
+	// in a goroutine and call subCancel() to exit early??
+	// that goroutine will act as watchdog, and cancel the work as early as possible??
+	// as a side-note: it can keep hammering the lock so maybe not a good idea??
+
+	executedReqs := 0
 	for _, req := range reqs {
+		// if shouldQuit is true, terminate and abandon requests
 		if overallResponse.shouldQuit() {
+			// cancel currently running goroutines
+			subCancel()
```

Comment:
instead of doing this. what do you think about passing the `subCancel` func into the searchResponse object? Then, everytime `addRequest` or `setError` or `setStatus` is called you can check at the end of that function and call `subCancel` if necessary.

i.e. the searchResponse object can be responsible for cancelling the context. it would also cancel slightly quicker when the search limit is reached.

Reply:
makes sense, passing around cancelFunc, and cancelling it in each event on searchResponse.